Nos/Views/CreateUsernameSheet.swift

```diff
@@ -0,0 +1,454 @@
+                .disabled(verified != true || isVerifying || invalidInput)
+            }
+        }
+        .sheetPage()
+    }
+
+    private func usernameAlreadyClaimedText() -> some View {
+        PlainText(.localizable.usernameAlreadyClaimed)
+            .font(.clarity(.medium, textStyle: .subheadline))
+            .foregroundStyle(Color.red)
+            .lineSpacing(3)
+    }
+
+    private var invalidInput: Bool {
+        usernameObserver.text.count < 3
+    }
+
+    private var validationFailed: Bool {
+        verified == false
+    }
+
+    private func verify(_ username: String) async {
+        verified = nil
+
+        guard !username.isEmpty, let keyPair = currentUser.keyPair else {
+            return
+        }
+
+        isVerifying = true
+
+        defer {
+            isVerifying = false
+        }
+
+        do {
+            verified = try await namesAPI.verify(username: username, keyPair: keyPair)
+        } catch {
+            Log.error(error.localizedDescription)
+        }
+    }
+}
+
+fileprivate struct UsernameTextField: View {
+
+    @StateObject var usernameObserver: UsernameObserver
+    @FocusState private var usernameFieldIsFocused: Bool
+
+    var body: some View {
+        SwiftUI.TextField(
+            text: $usernameObserver.text,
+            prompt: PlainText(.localizable.username).foregroundStyle(Color.secondaryTxt)
+        ) {
+            PlainText(.localizable.username)
+                .foregroundStyle(Color.primaryTxt)
+        }
+        .focused($usernameFieldIsFocused)
+        .font(.clarity(.bold, textStyle: .title3))
+        .textInputAutocapitalization(.never)
+        .textCase(.lowercase)
+        .autocorrectionDisabled()
+        .foregroundStyle(Color.primaryTxt)
+        .lineLimit(1)
+        .padding(10)
+        .cornerRadius(10)
+        .overlay {
+            RoundedRectangle(cornerRadius: 10)
+                .stroke(Color.secondaryTxt, lineWidth: 2)
+        }
+        .background {
+            Color.black.opacity(0.1).cornerRadius(10)
+        }
+        .onChange(of: usernameObserver.text) { oldValue, newValue in
+            let characterset = CharacterSet(
+                charactersIn: "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789-."
+            )
+            if newValue.rangeOfCharacter(from: characterset.inverted) != nil {
+                usernameObserver.text = oldValue
+            } else if newValue.count > 30 {
+                usernameObserver.text = oldValue
+            } else {
+                usernameObserver.text = newValue.lowercased()
+            }
+        }
+    }
+}
+
+fileprivate struct ExcellentChoicePage: View {
+
+    var username: String
+    @Binding var isPresented: Bool
+    @State private var claimState: ClaimState = .idle
+    @Dependency(\.currentUser) var currentUser
+    @Dependency(\.namesAPI) var namesAPI
+
+    /// The current state of the claim request.
+    private enum ClaimState {
```

Comment:
Looks like my suggestion to create this enum made some of the other code more complex. Sorry about that -- didn't realize it would do that.